### DISCUSSION TITLE
oncePer image tag (version)

### DISCUSSION THREAD

**Discussion by lomboboo** [argoproj/argo-cd#11870] [General]
Hello, I am currently struggling with setting up notification once per image tag or app version. 
There is only an example in the docs with `oncePer: app.status.operationState.syncResult.revision`, but for my case I only want to send notification when the app version changes itself rather than every change in yaml configs.

The `app` object has only one path that contains what I need - `app.status.summary.images`, but it is a list of container images which can vary. 

If `oncePer` was accepting value instead of path I would be able to achieve what I need, but I guess it is not possible (any links to the source code of how it is implemented maybe?)

My question is how can I make argocd notification send only on image tag change or app version? Is there any standards to do this the right way?

**Comment by bcbrockway** [comment]
Yeah revision doesn't work for us for this reason _and_ because we have manifests defined for a number of different apps in the same repo; when _one_ of them is updated, _all_ of them get synced. I've got our template filtering out the correct image nicely now using the Image Updater annotations:

```yaml
  template.mixpanel-annotation-app-version: |
    webhook:
      mixpanel:
        method: POST
        path: app/projects/{{ .recipient }}/annotations
        body: |
          {{- $match := "" -}}
          {{- range $deployedImage := .app.status.summary.images -}}
          {{- range $configuredImage := split "," (index $.app.metadata.annotations "argocd-image-updater.argoproj.io/image-list") -}}
          {{- $configuredImage := (regexReplaceAll "(.*)(:.*)$" (index (split "=" $configuredImage) "_1") "${1}") -}}
          {{- if contains $configuredImage $deployedImage -}}
          {{- $match = $deployedImage -}}
          {{- end -}}
          {{- end -}}
          {{- end -}}
          {
            "date": "{{ (call .time.Now).Local.Format "2006-01-02 15:04:05" }}",
            "description": "MY_EKS_CLUSTER | {{ base $match }}"
          }
```

I can't figure out a way to get to it trigger only when one of the images in the list you mentioned changes though :confused:

**Comment by ggerogery** [comment]
`app.status.summary.images` - in my case I have fixed number of images in Application, just need to know an index with my image which we can find with this template: `{{ index .app.status.summary.images 2 }}` . Here is my template:

```
template.app-deployed: |
  message: |
    {{ $image_tag := (split ":" (index .app.status.summary.images 2))._1 }}
    {{ $commit_hash := (split "-" $image_tag)._1 }}
    `{{.app.metadata.name}}` environment has been updated with image `{{ $image_tag }}`.
    <https://bitbucket.org/company/product/commits/{{ $commit_hash }}|See Bitbucket commit>
```  

` {{ $image_tag := (split ":" (index .app.status.summary.images 2))._1 }}`
This template converts `xxxxxxxxx.dkr.ecr.us-east-1.amazonaws.com/backend/celery:some_image_tag` to just `some_image_tag`

Also, I found that most of [Helm functions](https://helm.sh/docs/chart_template_guide/function_list/) works in template

**Comment by iarc13** [comment]
> Hello, I am currently struggling with setting up notification once per image tag or app version. There is only an example in the docs with `oncePer: app.status.operationState.syncResult.revision`, but for my case I only want to send notification when the app version changes itself rather than every change in yaml configs.
> 
> The `app` object has only one path that contains what I need - `app.status.summary.images`, but it is a list of container images which can vary.
> 
> If `oncePer` was accepting value instead of path I would be able to achieve what I need, but I guess it is not possible (any links to the source code of how it is implemented maybe?)
> 
> My question is how can I make argocd notification send only on image tag change or app version? Is there any standards to do this the right way?

I had a similar situation wherein I was getting a lot of notifications. To reduce that, I resorted to `oncePer`. I had multiple images, was using argo image updater. So what I did was:
```
    trigger.on-deployed: |
      - description: Application is synced and healthy.
        oncePer: app.spec.source.kustomize.images
        send:
        - app-deployed
        when: app.status.operationState.phase in ['Succeeded'] and app.status.health.status == 'Healthy'
```

And it seems to work. Argo can take image list and remember that. The only problem I face is that quite often I get the image tag of the last one as summary is updated a bit late I guess. So I'll try to fetch image tag from `.app.spec.source.kustomize.images` instead.

**Reply by lomboboo** [reply]
We ended up using similar approach, but Helm properties are different from Kustomize, so they end up with different configuration. Anyways, I believe it shouldn't be that hard to configure and maybe ArgoCD team will eventually fix it to be easy to use. For now, it's a lot of hassle to setup and it doesn't always work 100% as intended.
